File: backend/src/main/java/com/aetheris/rag/service/RecommendationService.java
/*
 * Copyright 2025 Aetheris RAG Team. All rights reserved.
 */
package com.aetheris.rag.service;

import com.aetheris.rag.dto.response.RecommendationResponse;

/**
 * 推荐服务接口。
 *
 * <p>提供基于用户画像的个性化推荐功能，支持：
 * <ul>
 *   <li>个性化推荐：基于用户画像向量检索相关资源</li>
 *   <li>默认推荐：无画像时按热度/时间返回资源</li>
 *   <li>推荐理由和学习建议生成（通过 LLM）</li>
 *   <li>证据引用（Citations）</li>
 * </ul>
 *
 * @author Aetheris Team
 * @version 1.0.0
 * @since 2026-03-28
 */
public interface RecommendationService {

  /**
   * 为用户生成个性化推荐。
   *
   * <p>推荐流程：
   * <ol>
   *   <li>获取用户画像向量（如果有）</li>
   *   <li>使用画像向量进行 KNN 检索（或使用默认策略）</li>
   *   <li>按资源聚合检索结果</li>
   *   <li>调用 LLM 生成推荐理由和学习建议</li>
   *   <li>构建推荐响应（包含引用证据）</li>
   * </ol>
   *
   * @param userId 用户ID
   * @param topN 推荐数量（1-50）
   * @return 推荐响应
   */
  RecommendationResponse recommend(Long userId, int topN);

  /**
   * 异步记录推荐点击行为并更新画像。
   *
   * @param userId 用户ID
   * @param resourceId 资源ID
   */
  void recordClickAsync(Long userId, Long resourceId);

  /**
   * 异步记录推荐收藏行为并更新画像。
   *
   * @param userId 用户ID
   * @param resourceId 资源ID
   */
  void recordFavoriteAsync(Long userId, Long resourceId);
}

File: frontend/src/services/recommendation.service.ts
/*
 * Copyright 2025 Aetheris RAG Team. All rights reserved.
 */

import api from './api'

/**
 * 推荐资源项接口
 */
export interface RecommendationItem {
  /** 资源ID */
  resourceId: string
  /** 资源标题 */
  title: string
  /** 标签（逗号分隔） */
  tags: string
  /** AI 生成的推荐理由 */
  reason: string
  /** AI 生成的学习建议 */
  suggestion: string
  /** 证据引用列表 */
  citations: Citation[]
  /** 相似度分数 */
  score: number
  /** 文件类型 */
  fileType: string
  /** 资源描述 */
  description: string
}

/**
 * 引用对象（复用 chat.service.ts 中的 Citation 结构）
 */
export interface Citation {
  resourceId: string
  resourceTitle: string
  chunkId: string
  chunkIndex: number
  location: CitationLocation
  snippet: string
  score: number
}

/**
 * 引用位置信息
 */
export type CitationLocation = PdfLocation | MarkdownLocation

/**
 * PDF 位置信息
 */
export interface PdfLocation {
  type: 'pdf'
  pageStart: number
  pageEnd: number
}

/**
 * Markdown 位置信息
 */
export interface MarkdownLocation {
  type: 'markdown'
  chapterPath: string
}

/**
 * 推荐响应接口
 */
export interface RecommendationResponse {
  /** 推荐资源项列表 */
  items: RecommendationItem[]
  /** 是否为个性化推荐 */
  personalized: boolean
  /** 推荐生成耗时（毫秒） */
  latencyMs: number
  /** 推荐结果数量 */
  count: number
}

/**
 * 推荐服务类
 *
 * <p>封装推荐相关的 API 调用，包括：
 * <ul>
 *   <li>获取推荐列表</li>
 *   <li>记录推荐点击行为</li>
 *   <li>记录推荐收藏行为</li>
 * </ul>
 */
export class RecommendationService {
  /**
   * 获取个性化推荐列表
   *
   * @param topN 推荐数量（默认 10）
   * @returns 推荐响应
   */
  static async getRecommendations(topN = 10): Promise<RecommendationResponse> {
    const response = await api.get<any, RecommendationResponse>('/recommendations', {
      params: { topN }
    })
    return response
  }

  /**
   * 记录推荐资源点击行为
   *
   * @param resourceId 资源ID
   */
  static async recordClick(resourceId: string): Promise<void> {
    await api.post('/recommendations/click', null, {
      params: { resourceId }
    })
  }

  /**
   * 记录推荐资源收藏行为
   *
   * @param resourceId 资源ID
   */
  static async recordFavorite(resourceId: string): Promise<void> {
    await api.post('/recommendations/favorite', null, {
      params: { resourceId }
    })
  }
}

export default RecommendationService
